File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseRepository.java
package br.com.jonathanzanella.myexpenses.expense;

import android.os.AsyncTask;
import android.support.annotation.UiThread;
import android.support.annotation.WorkerThread;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;

import br.com.jonathanzanella.myexpenses.Environment;
import br.com.jonathanzanella.myexpenses.MyApplication;
import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.card.Card;
import br.com.jonathanzanella.myexpenses.card.CardRepository;
import br.com.jonathanzanella.myexpenses.chargeable.ChargeableType;
import br.com.jonathanzanella.myexpenses.database.Fields;
import br.com.jonathanzanella.myexpenses.database.ModelRepository;
import br.com.jonathanzanella.myexpenses.database.Repository;
import br.com.jonathanzanella.myexpenses.database.Where;
import br.com.jonathanzanella.myexpenses.helpers.DateHelper;
import br.com.jonathanzanella.myexpenses.overview.WeeklyPagerAdapter;
import br.com.jonathanzanella.myexpenses.validations.OperationResult;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;

import static br.com.jonathanzanella.myexpenses.chargeable.ChargeableType.ACCOUNT;
import static br.com.jonathanzanella.myexpenses.chargeable.ChargeableType.DEBIT_CARD;
import static br.com.jonathanzanella.myexpenses.helpers.DateHelper.firstDayOfMonth;
import static br.com.jonathanzanella.myexpenses.helpers.DateHelper.lastDayOfMonth;
import static br.com.jonathanzanella.myexpenses.log.Log.warning;

public class ExpenseRepository implements ModelRepository<Expense> {
	private Repository<Expense> repository;
	private CardRepository cardRepository;
	private ExpenseTable table = new ExpenseTable();

	public ExpenseRepository(Repository<Expense> repository) {
		this.repository = repository;
	}

	private CardRepository getCardRepository() {
		if(cardRepository == null)
			cardRepository = new CardRepository(new Repository<Card>(MyApplication.getContext()), this);
		return cardRepository;
	}

	@WorkerThread
	public Expense find(String uuid) {
		return repository.find(table, uuid);
	}

	@WorkerThread
	List<Expense> userExpenses() {
		return repository.query(table, new Where(Fields.USER_UUID).eq(Environment.CURRENT_USER_UUID).orderBy(Fields.DATE));
	}

	@WorkerThread
	public List<Expense> monthly(DateTime month) {
		DateTime lastMonth = month.minusMonths(1);
		DateTime initOfMonth = firstDayOfMonth(lastMonth);
		DateTime endOfMonth = lastDayOfMonth(lastMonth);

		List<Expense> expenses = repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
							.and(Fields.CHARGE_NEXT_MONTH).eq(true)
							.orderBy(Fields.DATE));

		initOfMonth = firstDayOfMonth(month);
		endOfMonth = lastDayOfMonth(month);

		expenses.addAll(repository.query(table, queryBetween(initOfMonth, endOfMonth)
				.and(Fields.REMOVED).eq(false)
				.and(Fields.CHARGE_NEXT_MONTH).eq(false)
				.and(Fields.USER_UUID).eq(Environment.CURRENT_USER_UUID)
				.orderBy(Fields.DATE)));

		return expenses;
	}
	public List<Expense> expenses(WeeklyPagerAdapter.Period period) {
		return expenses(period, null);
	}

	@WorkerThread
	public List<Expense> expenses(WeeklyPagerAdapter.Period period, Card card) {
		List<Expense> expenses = new ArrayList<>();

		if(period.init.getDayOfMonth() == 1) {
			DateTime date = DateHelper.firstDayOfMonth(period.init);
			DateTime initOfMonth = date.minusMonths(1);
			DateTime endOfMonth = DateHelper.lastDayOfMonth(initOfMonth);

			Where where = queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
					.and(Fields.CHARGE_NEXT_MONTH).eq(true)
					.and(Fields.IGNORE_IN_OVERVIEW).eq(false)
					.orderBy(Fields.DATE);
			if(card != null)
				where = where.and(Fields.CHARGEABLE_UUID).eq(card.getUuid());
			expenses.addAll(repository.query(table, where));
		}

		DateTime init = DateHelper.firstMillisOfDay(period.init);
		DateTime end = DateHelper.lastMillisOfDay(period.end);

		Where where = queryBetweenUserDataAndNotRemoved(init, end)
				.and(Fields.CHARGE_NEXT_MONTH).eq(false)
				.and(Fields.IGNORE_IN_OVERVIEW).eq(false)
				.orderBy(Fields.DATE);
		if(card != null)
			where = where.and(Fields.CHARGEABLE_UUID).eq(card.getUuid());
		expenses.addAll(repository.query(table, where));

		return expenses;
	}

	@WorkerThread
	public List<Expense> expensesForResumeScreen(DateTime date) {
		DateTime lastMonth = date.minusMonths(1);
		DateTime initOfMonth = DateHelper.firstDayOfMonth(lastMonth);
		DateTime endOfMonth = DateHelper.lastDayOfMonth(lastMonth);

		List<Expense> expenses = repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
				.and(Fields.CHARGEABLE_TYPE).notEq(ChargeableType.CREDIT_CARD.name())
				.and(Fields.CHARGE_NEXT_MONTH).eq(true)
				.and(Fields.IGNORE_IN_OVERVIEW).eq(false)
				.orderBy(Fields.DATE));

		initOfMonth = DateHelper.firstDayOfMonth(date);
		endOfMonth = DateHelper.lastDayOfMonth(date);

		expenses.addAll(repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
				.and(Fields.CHARGEABLE_TYPE).notEq(ChargeableType.CREDIT_CARD.name())
				.and(Fields.CHARGE_NEXT_MONTH).eq(false)
				.and(Fields.IGNORE_IN_OVERVIEW).eq(false)
				.orderBy(Fields.DATE)));

		DateTime creditCardMonth = date.minusMonths(1);
		List<Card> cards = getCardRepository().creditCards();
		for (Card card : cards) {
			int total = getCardRepository().getInvoiceValue(card, creditCardMonth);
			if(total == 0)
				continue;

			Expense expense = new Expense();
			expense.setChargeable(card);
			expense.setName(MyApplication.getContext().getString(R.string.invoice));
			expense.setDate(creditCardMonth);
			expense.setValue(total);
			expense.setCreditCard(card);
			expenses.add(expense);
		}

		return expenses;
	}

	@WorkerThread
	public List<Expense> accountExpenses(Account account, DateTime month) {
		DateTime lastMonth = month.minusMonths(1);
		DateTime initOfMonth = DateHelper.firstDayOfMonth(lastMonth);
		DateTime endOfMonth = DateHelper.lastDayOfMonth(lastMonth);

		Card card = getCardRepository().accountDebitCard(account);

		List<Expense> expenses = repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
				.and(Fields.CHARGEABLE_TYPE).eq(ACCOUNT.name())
				.and(Fields.CHARGEABLE_UUID).eq(account.getUuid())
				.and(Fields.CHARGE_NEXT_MONTH).eq(true)
				.orderBy(Fields.DATE));

		if(card != null) {
			expenses.addAll(repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
					.and(Fields.CHARGEABLE_TYPE).eq(DEBIT_CARD.name())
					.and(Fields.CHARGEABLE_UUID).eq(card.getUuid())
					.and(Fields.CHARGE_NEXT_MONTH).eq(true)
					.orderBy(Fields.DATE)));
		}

		initOfMonth = firstDayOfMonth(month);
		endOfMonth = lastDayOfMonth(month);

		expenses.addAll(repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
				.and(Fields.CHARGEABLE_TYPE).eq(ACCOUNT.name())
				.and(Fields.CHARGEABLE_UUID).eq(account.getUuid())
				.and(Fields.CHARGE_NEXT_MONTH).eq(false)
				.orderBy(Fields.DATE)));

		if(card != null) {
			expenses.addAll(repository.query(table, queryBetweenUserDataAndNotRemoved(initOfMonth, endOfMonth)
					.and(Fields.CHARGEABLE_TYPE).eq(DEBIT_CARD.name())
					.and(Fields.CHARGEABLE_UUID).eq(card.getUuid())
					.and(Fields.CHARGE_NEXT_MONTH).eq(false)
					.orderBy(Fields.DATE)));
		}

		if(account.isAccountToPayCreditCard()) {
			DateTime creditCardMonth = month.minusMonths(1);
			List<Card> cards = getCardRepository().creditCards();
			for (Card creditCard : cards) {
				int total = getCardRepository().getInvoiceValue(card, creditCardMonth);
				if (total == 0)
					continue;

				Expense expense = new Expense();
				expense.setChargeable(card);
				expense.setName(MyApplication.getContext().getString(R.string.invoice) + " " + creditCard.getName());
				expense.setDate(creditCardMonth.plusMonths(1));
				expense.setValue(total);
				expense.setCreditCard(card);
				expenses.add(expense);
			}
		}

		Collections.sort(expenses, new Comparator<Expense>() {
			@Override
			public int compare(Expense lhs, Expense rhs) {
				if(lhs.getDate().isAfter(rhs.getDate()))
					return 1;
				return -1;
			}
		});

		return expenses;
	}

	@WorkerThread
	public long greaterUpdatedAt() {
		return repository.greaterUpdatedAt(table);
	}

	@WorkerThread
	public List<Expense> unsync() {
		return repository.unsync(table);
	}

	private Where queryBetween(DateTime init, DateTime end) {
		return new Where(Fields.DATE).greaterThanOrEq(init.getMillis())
				.and(Fields.DATE).lessThanOrEq(end.getMillis());
	}

	private Where queryBetweenUserDataAndNotRemoved(DateTime init, DateTime end) {
		return queryBetween(init, end)
				.and(Fields.REMOVED).eq(false)
				.and(Fields.USER_UUID).eq(Environment.CURRENT_USER_UUID);
	}

	@WorkerThread
	public OperationResult save(Expense expense) {
		OperationResult result = new OperationResult();
		if(StringUtils.isEmpty(expense.getName()))
			result.addError(ValidationError.NAME);
		if(expense.getValue() <= 0)
			result.addError(ValidationError.AMOUNT);
		if(expense.getDate() == null)
			result.addError(ValidationError.DATE);
		if(expense.getChargeable() == null)
			result.addError(ValidationError.CHARGEABLE);
		if(result.isValid()) {
			if(expense.getId() == 0 && expense.getUuid() == null)
				expense.setUuid(UUID.randomUUID().toString());
			if(expense.getId() == 0 && expense.getUserUuid() == null)
				expense.setUserUuid(Environment.CURRENT_USER_UUID);
			expense.setSync(false);
			repository.saveAtDatabase(table, expense);
		}
		return result;
	}

	@UiThread
	void saveAsync(final Expense expense) {
		new AsyncTask<Void, Void, OperationResult>() {

			@Override
			protected OperationResult doInBackground(Void... voids) {
				return save(expense);
			}

			@Override
			protected void onPostExecute(OperationResult operationResult) {
				super.onPostExecute(operationResult);
				if(!operationResult.isValid())
					throw new UnsupportedOperationException("Could not save expense " + expense.getUuid());
			}
		}.execute();
	}

	@WorkerThread
	@Override
	public void syncAndSave(final Expense unsyncExpense) {
		Expense expense = find(unsyncExpense.getUuid());
		if(expense != null && expense.getId() != unsyncExpense.getId()) {
			if(expense.getUpdatedAt() != unsyncExpense.getUpdatedAt())
				warning("Expense overwritten", unsyncExpense.getData());
			unsyncExpense.setId(expense.getId());
		}

		unsyncExpense.setSync(true);
		repository.saveAtDatabase(table, unsyncExpense);
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseRepository.kt
package br.com.jonathanzanella.myexpenses.expense

import android.support.annotation.WorkerThread
import android.util.Log
import br.com.jonathanzanella.myexpenses.MyApplication
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.card.Card
import br.com.jonathanzanella.myexpenses.card.CardRepository
import br.com.jonathanzanella.myexpenses.chargeable.ChargeableType
import br.com.jonathanzanella.myexpenses.helpers.DateHelper
import br.com.jonathanzanella.myexpenses.helpers.DateHelper.firstDayOfMonth
import br.com.jonathanzanella.myexpenses.helpers.DateHelper.lastDayOfMonth
import br.com.jonathanzanella.myexpenses.overview.WeeklyPagerAdapter
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.validations.ValidationResult
import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTime
import java.util.*

open class ExpenseRepository() {
    private val cardRepository: CardRepository by lazy {
        CardRepository(this)
    }

    @WorkerThread
    fun find(uuid: String): Expense? {
        return MyApplication.database.expenseDao().find(uuid).blockingFirst().firstOrNull()
    }

    @WorkerThread
    fun all(): List<Expense> {
        return MyApplication.database.expenseDao().all().blockingFirst()
    }

    @WorkerThread
    fun monthly(month: DateTime): List<Expense> {
        val lastMonth = month.minusMonths(1)
        var initOfMonth = firstDayOfMonth(lastMonth)
        var endOfMonth = lastDayOfMonth(lastMonth)

        val expenses = MyApplication.database.expenseDao().nextMonth(initOfMonth.millis, endOfMonth.millis).blockingFirst()

        initOfMonth = firstDayOfMonth(month)
        endOfMonth = lastDayOfMonth(month)

        expenses.addAll(MyApplication.database.expenseDao().currentMonth(initOfMonth.millis, endOfMonth.millis).blockingFirst())

        return expenses
    }

    fun expenses(period: WeeklyPagerAdapter.Period): List<Expense> {
        return expenses(period, null)
    }

    @WorkerThread
    fun expenses(period: WeeklyPagerAdapter.Period, card: Card?): List<Expense> {
        val expenses = ArrayList<Expense>()

        if (period.init?.dayOfMonth == 1) {
            val date = DateHelper.firstDayOfMonth(period.init!!)
            val initOfMonth = date.minusMonths(1)
            val endOfMonth = DateHelper.lastDayOfMonth(initOfMonth)

            if (card != null)
                expenses.addAll(MyApplication.database.expenseDao().overviewNextMonth(initOfMonth.millis, endOfMonth.millis, card.uuid!!).blockingFirst())
            else
                expenses.addAll(MyApplication.database.expenseDao().overviewNextMonth(initOfMonth.millis, endOfMonth.millis).blockingFirst())
        }

        val init = DateHelper.firstMillisOfDay(period.init!!)
        val end = DateHelper.lastMillisOfDay(period.end!!)

        if (card != null)
            expenses.addAll(MyApplication.database.expenseDao().overviewCurrentMonth(init.millis, end.millis, card.uuid!!).blockingFirst())
        else
            expenses.addAll(MyApplication.database.expenseDao().overviewCurrentMonth(init.millis, end.millis).blockingFirst())

        return expenses
    }

    @WorkerThread
    fun unpaidCardExpenses(month: DateTime, card: Card): List<Expense> {
        val expenses = ArrayList<Expense>()
        val lastMonth = month.minusMonths(1)

        expenses.addAll(MyApplication.database.expenseDao().unchargedNextMonth(firstDayOfMonth(lastMonth).millis, lastDayOfMonth(lastMonth).millis, card.uuid!!).blockingFirst())
        expenses.addAll(MyApplication.database.expenseDao().unchargedCurrentMonth(firstDayOfMonth(lastMonth).millis, lastDayOfMonth(lastMonth).millis, card.uuid!!).blockingFirst())

        return expenses
    }

    @WorkerThread
    fun expensesForResumeScreen(date: DateTime): List<Expense> {
        val lastMonth = date.minusMonths(1)
        var initOfMonth = DateHelper.firstDayOfMonth(lastMonth)
        var endOfMonth = DateHelper.lastDayOfMonth(lastMonth)

        val expenses = MyApplication.database.expenseDao().resumeNextMonth(initOfMonth.millis, endOfMonth.millis, ChargeableType.CREDIT_CARD.name).blockingFirst()

        initOfMonth = DateHelper.firstDayOfMonth(date)
        endOfMonth = DateHelper.lastDayOfMonth(date)

        expenses.addAll(MyApplication.database.expenseDao().resumeCurrentMonth(initOfMonth.millis, endOfMonth.millis, ChargeableType.CREDIT_CARD.name).blockingFirst())

        val creditCardMonth = date.minusMonths(1)
        val cards = cardRepository.creditCards()
        for (card in cards) {
            val total = cardRepository.getInvoiceValue(card, creditCardMonth)
            if (total == 0)
                continue

            val expense = Expense()
            expense.setChargeable(card)
            expense.name = MyApplication.getContext().getString(R.string.invoice)
            expense.setDate(creditCardMonth)
            expense.value = total
            expense.creditCard = card
            expenses.add(expense)
        }

        return expenses
    }

    @WorkerThread
    fun accountExpenses(account: Account, month: DateTime): List<Expense> {
        val lastMonth = month.minusMonths(1)
        var initOfMonth = DateHelper.firstDayOfMonth(lastMonth)
        var endOfMonth = DateHelper.lastDayOfMonth(lastMonth)

        val card = cardRepository.accountDebitCard(account)

        val expenses = MyApplication.database.expenseDao().nextMonth(initOfMonth.millis, endOfMonth.millis, account.uuid!!).blockingFirst()

        if (card != null)
            expenses.addAll(MyApplication.database.expenseDao().nextMonth(initOfMonth.millis, endOfMonth.millis, card.uuid!!).blockingFirst())

        initOfMonth = firstDayOfMonth(month)
        endOfMonth = lastDayOfMonth(month)

        expenses.addAll(MyApplication.database.expenseDao().currentMonth(initOfMonth.millis, endOfMonth.millis, account.uuid!!).blockingFirst())

        if (card != null)
            expenses.addAll(MyApplication.database.expenseDao().currentMonth(initOfMonth.millis, endOfMonth.millis, card.uuid!!).blockingFirst())

        if (account.isAccountToPayCreditCard) {
            val creditCardMonth = month.minusMonths(1)
            val cards = cardRepository.creditCards()
            for (creditCard in cards) {
                val total = cardRepository.getInvoiceValue(creditCard, creditCardMonth)
                if (total == 0)
                    continue

                val expense = Expense()
                expense.setChargeable(creditCard)
                expense.name = MyApplication.getContext().getString(R.string.invoice) + " " + creditCard.name
                expense.setDate(creditCardMonth.plusMonths(1))
                expense.value = total
                expense.creditCard = creditCard
                expenses.add(expense)
            }
        }

        Collections.sort(expenses, Comparator<Expense> { lhs, rhs ->
            if (lhs.getDate().isAfter(rhs.getDate()))
                return@Comparator 1
            -1
        })

        return expenses
    }

    @WorkerThread
    fun greaterUpdatedAt(): Long {
        return MyApplication.database.expenseDao().greaterUpdatedAt().blockingFirst().firstOrNull()?.updatedAt ?: 0L
    }

    @WorkerThread
    fun unsync(): List<Expense> {
        return MyApplication.database.expenseDao().unsync().blockingFirst()
    }

    @WorkerThread
    fun save(expense: Expense): ValidationResult {
        val result = validate(expense)
        if (result.isValid) {
            if (expense.id == 0L && expense.uuid == null)
                expense.uuid = UUID.randomUUID().toString()
            expense.sync = false
            expense.id = MyApplication.database.expenseDao().saveAtDatabase(expense)
        }
        return result
    }

    private fun validate(expense: Expense): ValidationResult {
        val result = ValidationResult()
        if (StringUtils.isEmpty(expense.name))
            result.addError(ValidationError.NAME)
        if (expense.value == 0)
            result.addError(ValidationError.AMOUNT)
        if (!expense.dateIsPresent())
            result.addError(ValidationError.DATE)
        if (expense.chargeableFromCache == null)
            result.addError(ValidationError.CHARGEABLE)
        return result
    }

    @WorkerThread
    fun syncAndSave(unsync: Expense): ValidationResult {
        val result = validate(unsync)
        if (!result.isValid) {
            Log.w("Expense validation fail", unsync.getData() + "\nerrors: " + result.errorsAsString)
            return result
        }

        val expense = find(unsync.uuid!!)
        if (expense != null && expense.id != unsync.id) {
            if (expense.updatedAt != unsync.updatedAt)
                Log.w("Expense overwritten", unsync.getData())
            unsync.id = expense.id
        }

        unsync.sync = true
        unsync.id = MyApplication.database.expenseDao().saveAtDatabase(unsync)

        return result
    }
}